on-chain/evm-contracts/contracts/Bridge.sol

```diff
@@ -0,0 +1,387 @@
+pragma solidity ^0.5.12;
+
+import "./helpers/SafeMath.sol";
+import "./interfaces/IValidator.sol";
+import "./interfaces/IERC20Handler.sol";
+import "./interfaces/IERC721Handler.sol";
+import "./interfaces/IDepositHandler.sol";
+
+contract Bridge {
+    using SafeMath for uint;
+
+    IValidator public _validatorContract;
+    uint public _validatorThreshold;
+    ValidatorThresholdProposal public _currentValidatorThresholdProposal;
+
+    enum Vote {No, Yes}
+
+    // ValidatorThresholdProposalStatus and _validatorThresholdProposalStatusStrings must be kept
+    // the same length and order to function properly
+    enum ValidatorThresholdProposalStatus {Inactive, Active}
+    string[] _validatorThresholdProposalStatusStrings = ["inactive", "active"];
+
+    // DepositProposalStatus and _depositProposalStatusStrings must be kept
+    // the same length and order to function properly
+    enum DepositProposalStatus {Inactive, Active, Denied, Passed, Transferred}
+    string[] _depositProposalStatusStrings = ["inactive", "active", "denied", "passed", "transferred"];
+
+    struct GenericDepositRecord {
+        address _originChainTokenAddress;
+        address _originChainHandlerAddress;
+        uint    _destinationChainID;
+        address _destinationChainHandlerAddress;
+        address _destinationRecipientAddress;
+        bytes   _data;
+    }
+
+    struct ERC20DepositRecord {
+        address _originChainTokenAddress;
+        address _originChainHandlerAddress;
+        uint    _destinationChainID;
+        address _destinationChainHandlerAddress;
+        address _destinationRecipientAddress;
+        uint    _amount;
+    }
+
+    struct ERC721DepositRecord {
+        address _originChainTokenAddress;
+        address _originChainHandlerAddress;
+        uint    _destinationChainID;
+        address _destinationChainHandlerAddress;
+        address _destinationRecipientAddress;
+        uint    _tokenID;
+        bytes   _data;
+    }
```

Comment:
What if these were moved to the respective handlers?

We have a generic deposit function that calls the respective handler, and then we emit the deposit event from within this contract?

Reply:
I think it might be a good idea to do this if the generic deposit method in the Bridge contract accepts an arbitrary bytes argument that would be unpacked within the handler contracts. This way if any of the deposit record structs changed, you'd only have to redeploy that specific handler rather than the Bridge contract and the handler.

If we changed this, the `genericDeposit` method could look like this:
```solidity
function deposit(
        address      handlerAddress,
        bytes memory data
    ) public {
        uint depositID = ++_depositCounts[handlerAddress];

        _depositRecords[handlerAddress][depositID] = DepositRecord(
            handlerAddress,
            data
        );

        IDepositHandler depositHandler = IDepositHandler(handlerAddress);
        depositHandler.deposit(depositID, data);

        emit Deposit(depositID, handlerAddress);
    }
```

Reply:
i like it 
Please make an issue for it

Reply:
Here is the issue for reference #174